### PR TITLE
New UDF: Legacy compatible experiments 

### PR DESCRIPTION
**Pull request by rebecca-burwei**
Checklist for reviewer:

- [ ] Commits should reference a bug or github issue, if relevant (if a bug is referenced, the pull request should include the bug number in the title)
- [ ] If the PR comes from a fork, trigger integration CI tests by running the [Push to upstream workflow](https://github.com/mozilla/bigquery-etl/actions/workflows/push-to-upstream.yml) and provide the `<username>:<branch>` of the fork as parameter. The parameter will also show up
in the logs of the `manual-trigger-required-for-fork` CI task together with more detailed instructions.
- [ ] If adding a new field to a query, ensure that the schema and dependent downstream schemas have been updated

For modifications to schemas in restricted namespaces (see [`CODEOWNERS`](/CODEOWNERS)):
- [ ] Follow the [change control procedure](https://docs.google.com/document/d/1TTJi4ht7NuzX6BPG_KTr6omaZg70cEpxe9xlpfnHj9k/edit#heading=h.ttegrcfy18ck)

### CODE REVIEW

**Review by scholtzan — APPROVED** [review]
Looks good to me. The CI errors are unrelated to this change

File: sql/mozfun/glean/legacy_compatible_experiments/metadata.yaml [hunk 1]
```diff
@@ -0,0 +1,2 @@
+description: Formats a Glean experiments field into a Legacy Telemetry experiments field by dropping the extra information that Glean collects
```
Comment (scholtzan) [inline sql/mozfun/glean/legacy_compatible_experiments/metadata.yaml:1]:
```suggestion
description: >
  Formats a Glean experiments field into a Legacy Telemetry experiments field
  by dropping the extra information that Glean collects
```
Oh, for yaml files the max line width is 80 characters, so we need to split this across multiple lines

**Comment by rebecca-burwei** [comment]
> Looks good to me. The CI errors are unrelated to this change

Thanks @scholtzan . It seems like I can't merge in this PR with failing required tests though.

**Comment by scholtzan** [comment]
Ok, this is strange. I see other PRs successfully running these checks. I'll have a look in a bit

File: sql/mozfun/glean/legacy_compatible_experiments/udf.sql [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+expected AS (
+  SELECT
+    ARRAY(
+      STRUCT("experiment_a" AS key, "control" AS value),
+      STRUCT("experiment_b" AS key, "treatment" AS value),
+    ) AS experiments
+),
+)
+SELECT
+  assert.equals(expected, glean.legacy_compatible_experiments(ping_info.experiments))
```
Comment (scholtzan) [inline sql/mozfun/glean/legacy_compatible_experiments/udf.sql:34]: This should fix it:

```suggestion
CREATE OR REPLACE FUNCTION glean.legacy_compatible_experiments(
  ping_info__experiments ARRAY<
    STRUCT<key STRING, value STRUCT<branch STRING, extra STRUCT<type STRING>>>
  >
)
RETURNS ARRAY<STRUCT<key STRING, value STRING>> AS (
  ARRAY(SELECT STRUCT(x.key AS key, x.value.branch AS value) FROM UNNEST(ping_info__experiments) x)
);

-- Tests
WITH ping_info AS (
  SELECT
    [
      STRUCT(
        "experiment_a" AS key,
        STRUCT("control" AS branch, STRUCT("firefox" AS type) AS extra) AS value
      ),
      STRUCT(
        "experiment_b" AS key,
        STRUCT("treatment" AS branch, STRUCT("firefoxOS" AS type) AS extra) AS value
      )
    ] AS experiments
),
expected AS (
  SELECT
    [
      STRUCT("experiment_a" AS key, "control" AS value),
      STRUCT("experiment_b" AS key, "treatment" AS value)
    ] AS experiments
)
SELECT
  assert.equals(expected, glean.legacy_compatible_experiments(experiments))
FROM
  ping_info
```

File: sql/mozfun/glean/legacy_compatible_experiments/udf.sql [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+  SELECT
+    [
+      STRUCT("experiment_a" AS key, "control" AS value),
+      STRUCT("experiment_b" AS key, "treatment" AS value)
+    ] AS experiments
+)
+SELECT
+  assert.equals(expected, glean.legacy_compatible_experiments(experiments))
+FROM
+  ping_info
```
Comment (scholtzan) [inline sql/mozfun/glean/legacy_compatible_experiments/udf.sql:34]:
```suggestion
SELECT
  assert.equals(expected.experiments, glean.legacy_compatible_experiments(ping_info.experiments))
FROM
  ping_info,
  expected
```

File: sql/mozfun/glean/legacy_compatible_experiments/udf.sql [hunk 4]
```diff
@@ -0,0 +1,35 @@
...
+),
+expected AS (
+  SELECT
+    [
+      STRUCT("experiment_a" AS key, "control" AS value),
+      STRUCT("experiment_b" AS key, "treatment" AS value)
+    ] AS experiments
+)
+SELECT
+  assert.equals(expected.experiments, glean.legacy_compatible_experiments(ping_info.experiments))
```
Comment (scholtzan) [inline sql/mozfun/glean/legacy_compatible_experiments/udf.sql:32]:
```suggestion
  assert.array_equals(
    expected.experiments,
    glean.legacy_compatible_experiments(ping_info.experiments)
  )
```

**Comment by rebecca-burwei** [comment]
Hi @scholtzan - seems like one of the required tests is still failing

**Comment by scholtzan** [comment]
For some reason the test was timing out before, but succeeded now. The UDF should get automatically deployed by tomorrow.

**Comment by rebecca-burwei** [comment]
thank you @ascholtzan !

**Pull request merged**